[yashpola] iP

yashpola opened the pull request:
### Nicole Mark II
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Nicole Mark II is your easy-to-use task manager for important notables. Its, 

- text-based
- easy to learn
- ~FAST~ _SUPER_ FAST to use

All you need to do is, 

1. download it from [here](https://github.com/yashpola/ip/releases/tag/A-Jar)
2. Copy the jar file into an empty folder.
3. Open a command window in that folder.
4. Run the command java -jar nicole.jar (i.e., run the command in the same folder as the jar file).
5. Enjoy! 😉

And it is **FREE!**

Features: 
- [x] Managing todos, deadlines and events
- [x] Sorting by urgency
- [x] Searching by keyword
- [x] Saving between sessions

---
A little snippet of the the `greet()` function ;)
```java
private String greet() {
        DateTimeFormatter digitalTime = DateTimeFormatter.ofPattern("dd/MM/yyyy");
        DateTimeFormatter parseHour = DateTimeFormatter.ofPattern("HH");
        LocalDateTime now = LocalDateTime.now();
        String salutation = "Good ";
        int currentHour = Integer.parseInt(parseHour.format(now));
        if (currentHour < 12 && currentHour >= 5) {
            salutation += "morns :D ! What's the plan today";
        } else if (currentHour >= 12 && currentHour < 17) {
            salutation += "afternoon zzz...how's it going";
        } else {
            salutation += "evening! Let's wrap it up soon? :P";
        }
        return "Welcome to Nicole Mark II. New conversation at "
                + digitalTime.format(now) + "\n"
                + "----------------------------- x ------------------------------"
                + "\n" + "Nicole: " + salutation;
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

yashpola committed:
Level-0. Rename, greet, exit

yashpola committed:
Level-1. Echo

yashpola committed:
Level-2. Add, List

yashpola committed:
Level-3. Mark as Done

yashpola committed:
Level-4. ToDos, Events, Deadlines

yashpola committed:
Level-4. ToDos, Events, Deadlines

yashpola committed:
Level-4. ToDos, Events, Deadlines

yashpola committed:
Level-5. Automated Text UI Testing

yashpola committed:
Level-5. Handle Errors

yashpola committed:
Level-6. Delete

yashpola committed:
Nicole and Request class: add saving user tasks to hard drive

yashpola committed:
Nicole and Request class: add updating hard drive file for tasklist modifications

yashpola committed:
Nicole and Task classes: Add datetime parsing and priority ordering of tasks per datetime

yashpola committed:
Merge Level 8 with master

yashpola committed:
Add storage, ui, parser, tasklist classes

yashpola committed:
Increment: Add more oop

yashpola committed:
Move files into separate packages for better modularization

yashpola committed:
Merge branch-Packages into master

yashpola committed:
Merge gradle support

yashpola committed:
Configure build.gradle

yashpola committed:
Add JUnit tests for deadline, event, todo files

yashpola committed:
Merge branch-JUnit to master

yashpola committed:
Add JavaDocs for all non-trivial public methods

yashpola committed:
Introduce coding standard for all classes

yashpola committed:
Add find function based on keyword-match

yashpola committed:
Merge branch-A-JavaDoc to master

yashpola committed:
Merge branch-A-CodingStandard to master

yashpola committed:
Merge branch-Level-9 to master

yashpola committed:
Improve edge-case handling for mark, unmark, delete and find commands

yashpola committed:
Bug fixes for Level-7

yashpola committed:
Configure checkstyle and begin remedy

yashpola committed:
Add gui (javafx) to Nicole

yashpola committed:
Merge branch-Level-10 with master

taufiq commented on src/main/java/nicole/gui/MainWindow.java:
```diff
@@ -0,0 +1,97 @@
+            this.exit();
+        } else {
+            try {
+                Ui ui = new Ui();
+                String response = ui.talkToUser(input);
+                dialogContainer.getChildren().addAll(
+                        DialogBox.getUserDialog(input, USER_IMAGE),
+                        DialogBox.getNicoleDialog(response, NICOLE_IMAGE)
+                );
```
Hmm the dialog container can be part of the UI class

taufiq commented on src/main/java/nicole/task/Task.java:
```diff
@@ -0,0 +1,91 @@
+import nicole.nicoleexceptions.NicoleException;
+
+public class Task {
+    private boolean taskCompleted;
```
Boolean variables should be named with a "is","has","was" prefix (as stated by the Java conventions in CS2103)

taufiq commented on src/main/java/nicole/task/Task.java:
```diff
@@ -0,0 +1,91 @@
+     *                         invalid request.
+     */
+    public static Task taskFactory(String name, char taskType) throws NicoleException {
+        if (taskType == 'T') {
```
You could use a switch statement! This would then allow unknown `taskType` values that weren't captured in your cases to have a default handler flow.

taufiq commented on src/main/java/nicole/taskstorage/Storage.java:
```diff
@@ -0,0 +1,65 @@
+                numTasksInFile++;
+            }
+
+            Scanner userTaskFileReader = new Scanner(tasksFile);
```
Remember to close your scanner/readers

taufiq commented on src/main/java/nicole/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+        this.STORAGE = new Storage();
+    }
+
+    private void crudChecker(int taskNumber) throws NicoleException {
```
what is a crud checker? Would help if you could document it

taufiq reviewed (commented):
Looks sick! You might want to consider absorbing the JavaFX stuff into the UI class since that's what we'd expect a UI class to handle.

yashpola committed:
Extensions: Add method to update task name of any type, and method to check for datetime clashes for Event type tasks when adding new events

For the Update extension, it is necessary for user to be able to rectify mistakes in their tasks or simply update the descriptions if items change. For the Detecting Clashes extension, it is necessary for Nicole to be a coherent task manager and disallow obvious clashes in schedule.

For Update, rather than making users need to delete items to re-add them with the desired changes, we can allow them to change items directly

The Update extension requires a change of how task name is stored for Deadline and Event. Instead of storing the whole description including datetime, now only the description is stored and datetime is concatenated to their toString separately

yashpola committed:
Extensions: Add method to update task name of any type, and method to check for datetime clashes for Event type tasks when adding new events

For the Update extension, it is necessary for user to be able to rectify mistakes in their tasks or simply update the descriptions if items change. For the Detecting Clashes extension, it is necessary for Nicole to be a coherent task manager and disallow obvious clashes in schedule.

For Update, rather than making users need to delete items to re-add them with the desired changes, we can allow them to change items directly

The Update extension requires a change of how task name is stored for Deadline and Event. Instead of storing the whole description including datetime, now only the description is stored and datetime is concatenated to their toString separately

yashpola committed:
Extension: Add assertion to check for task creation.

There should not be a case where a Task type other than Event, Deadline or Todo is created. If there is an issue with String parsing when loading items from file and the task type is not correctly identified, there will be an assertion thrown.

yashpola committed:
Implement coding standard suggestions in PR, amend naming to be more descriptive, add comments in TaskList, Event, Deadline, and Storage explaining some non-obvious decisions, and improve coding standard in MainWindow userImage and nicoleImage fields to shorten line lengths.

yashpola committed:
Implement coding standard suggestions in PR, amend naming to be more descriptive, add comments in TaskList, Event, Deadline, and Storage explaining some non-obvious decisions, and improve coding standard in MainWindow userImage and nicoleImage fields to shorten line lengths.

yashpola committed:
Merge pull request #1 from yashpola/branch-A-CodeQuality

Branch A-CodeQuality

yashpola committed:
Revert "Implement coding standard suggestions in PR, amend naming to be more descriptive, add comments in TaskList, Event, Deadline, and Storage explaining some non-obvious decisions, and improve coding standard in MainWindow userImage and nicoleImage fields to shorten line lengths."

This reverts commit 3e59ea5d9d462b0ea0268ed40496afb62c2c7392.

yashpola committed:
Revert "Implement coding standard suggestions in PR, amend naming to be more descriptive, add comments in TaskList, Event, Deadline, and Storage explaining some non-obvious decisions, and improve coding standard in MainWindow userImage and nicoleImage fields to shorten line lengths."

This reverts commit 9ef9b152a9b39b2006ef6b2d3e1e6d80cb32b418.

yashpola committed:
Add assertion to null check tasks created from hard drive storage upon refreshing of chatbot

yashpola committed:
Merge pull request #2 from yashpola/branch-A-Assertions

Add assertion to null check tasks created from hard drive storage upo…

yashpola committed:
Implement coding standard suggestions in PR, amend naming to be more descriptive, add comments in TaskList, Event, Deadline, and Storage explaining some non-obvious decisions, and improve coding standard in MainWindow userImage and nicoleImage fields to shorten line lengths.

yashpola committed:
Merge pull request #3 from yashpola/branch-A-CodeQuality

Implement coding standard suggestions in PR, amend naming to be more …

tzaph commented on src/main/java/nicole/task/Deadline.java:
```diff
@@ -0,0 +1,68 @@
+        } catch (DateTimeParseException e) {
+            throw new NicoleException("Are you sure your date is in the proper [YYYY-MM-DD] format...?");
+        }
+        if (LocalDate.now().isAfter(deadlineDateLocalDate)) {
```
An interesting extension to the minimal criteria for the project!

tzaph commented on src/main/java/nicole/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+import nicole.task.Task;
+
+public class TaskList {
+    protected static final List<Task> TASKS = new ArrayList<>();
```
Is there a specific reason why we should use the keywords "static final" here? It is true that there can only be one list which makes it a constant, but the list is modifiable by adding and deleting elements, which makes the use of keywords a bit unintuitive to me.

tzaph commented on src/main/java/nicole/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    private void checkClashingTasks(Task newTask) throws NicoleException {
+        long numClashingTasks = TASKS.stream().filter(task -> {
```
I love the use of FP here!

tzaph commented on src/main/java/nicole/userrequests/Parser.java:
```diff
@@ -0,0 +1,47 @@
+     * @throws NicoleException if the request is unknown or in the wrong format.
+     */
+    protected static Task parseRequest(String name) throws NicoleException {
+        Pattern todoPattern = Pattern.compile("^todo(?: (.*?))?$");
```
Amazing use of regex here

tzaph reviewed (commented):
Incredible work on your iP so far! You exceeded the expected requirements and your coding style is neat and on point, which shows how experienced you are in writing Java code.

zoebelle-pang commented on src/main/java/nicole/task/Task.java:
```diff
@@ -0,0 +1,129 @@
+        this.isComplete = false;
+    }
+
+    /**
```
Good use of JavaDoc comments! 👍

zoebelle-pang commented on src/main/java/nicole/task/Todo.java:
```diff
@@ -0,0 +1,25 @@
+
+import nicole.nicoleexceptions.NicoleException;
+
+public class Todo extends Task {
```
Good job on using K&R style brackets throughout your code 😄

zoebelle-pang commented on src/main/java/nicole/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,196 @@
+        STORAGE.saveTasksToFile();
+        return unmarkedMessage;
+    }
+
```
Good use of white space throughout your code!

zoebelle-pang commented on src/main/java/nicole/taskstorage/TaskList.java:
```diff
@@ -0,0 +1,196 @@
+
+public class TaskList {
+    protected static final List<Task> TASKS = new ArrayList<>();
+    private static final Storage STORAGE = new Storage();
```
your code is generally well-organised 👍  since it contains a static final field named storage. just wondering if would it be more manageable in the future to modify the TaskList constructor to accept a Storage instance instead.

zoebelle-pang reviewed (commented):
Hello, 👋. Overall, your code follows Java coding standards and makes excellent use of whitespace, K&R style brackets, and comments. 😄 Good work!

yashpola committed:
Work on PR review comments, add JUnit tests, and fix bugs with storing Event descriptions and datetimes

yashpola committed:
Improve GUI and add product screenshot

yashpola committed:
Merge branch-A-BetterGui with master

yashpola committed:
Add user guide

yashpola committed:
Resize product ss in readme

yashpola committed:
Edit README

yashpola committed:
Chang SS for sort by date, and fix bug in sorting

yashpola committed:
Update readme

yashpola committed:
Update readme

yashpola committed:
Update readme

yashpola committed:
Update readme

yashpola committed:
Update readme and fix bug with file storage

yashpola committed:
Add back commented-out code in StorageTest

yashpola committed:
Update README

yashpola committed:
Update README, edit typo in help message

yashpola committed:
Update README, include additional exception in Event
